### ISSUE TITLE
Nondeterministic crashes from PackManager.retrieveAllBodies()

### ISSUE THREAD

**Issue by nrosner** [Sable/soot#1189]
Soot crashes nondeterministically on a 64-core machine when translating to Jimple.

When I repeatedly run the exact same Soot command-line invocation on the same jar file to translate to Jimple, it succeeds ~30% of the times, and fails ~70% of the times.

You can reproduce the problem by invoking Soot with `-process-dir` and translating a jar full of classes to Jimple.

Whenever it fails, it shows the error message:
```
[Thread-N] ERROR heros.solver.CountingThreadPoolExecutor - Worker thread execution failed: Failed to apply jb to <CLASS: METHOD>
```

Each time the command fails, the error message above displays a different method name from a different class, and a different thread number.

In the example below I'm using `-allow-phantom-refs` for simplicity, but the same problem occurs if you include the necessary classes in the Soot classpath. And I'm running it on a release jar from the Spring framework, but the same thing happens with almost any large jar file full of class files.

Example:
```
java -cp soot/target/sootclasses-trunk-jar-with-dependencies.jar soot.Main -pp -output-format jimple -allow-phantom-refs -process-dir spring-core-5.1.8.RELEASE.jar
```

There are five appearances of `availableProcessors()` in the Soot codebase. I tracked the issue down to the appearance on line 1261 of `src/main/java/soot/PackManager.java`:

```
    private void retrieveAllBodies() {                                                                                                                                                                                                                                                                                                                                                                                                      
      // The old coffi front-end is not thread-safe                                                                                                                                                                                                                                                                                                                                                                                         
      int threadNum = Options.v().coffi() ? 1 : Runtime.getRuntime().availableProcessors();                                                                                                                                                                                                                                                                                                                                            
```

If you replace that with a 1, Soot no longer crashes.

**Comment by nrosner** [comment]
Addendum: The underlying exception is
`Caused by: java.lang.RuntimeException: no active FastHierarchy present for scene`

Below is one example of a complete stack trace.

```
[Thread-26] ERROR heros.solver.CountingThreadPoolExecutor - Worker thread execution failed: Failed to apply jb to <org.springframework.core.io.buffer.DataBufferUtils$WritableByteChannelSubscriber: void hookOnNext(org.springframework.core.io.buffer.DataBuffer)>
java.lang.RuntimeException: Failed to apply jb to <org.springframework.core.io.buffer.DataBufferUtils$WritableByteChannelSubscriber: void hookOnNext(org.springframework.core.io.buffer.DataBuffer)>
	at soot.asm.AsmMethodSource.getBody(AsmMethodSource.java:2068)
	at soot.SootMethod.retrieveActiveBody(SootMethod.java:402)
	at soot.PackManager$1.run(PackManager.java:1279)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: no active FastHierarchy present for scene
	at soot.Scene.getFastHierarchy(Scene.java:1296)
	at soot.Scene.getOrMakeFastHierarchy(Scene.java:1287)
	at soot.toolkits.exceptions.ThrowableSet.whichCatchableAs(ThrowableSet.java:705)
	at soot.toolkits.graph.ExceptionalUnitGraph.buildExceptionDests(ExceptionalUnitGraph.java:290)
	at soot.toolkits.graph.ExceptionalUnitGraph.initialize(ExceptionalUnitGraph.java:229)
	at soot.toolkits.graph.ExceptionalUnitGraph.<init>(ExceptionalUnitGraph.java:135)
	at soot.toolkits.graph.ExceptionalUnitGraph.<init>(ExceptionalUnitGraph.java:163)
	at soot.jimple.toolkits.base.Aggregator.internalAggregate(Aggregator.java:129)
	at soot.jimple.toolkits.base.Aggregator.internalTransform(Aggregator.java:114)
	at soot.BodyTransformer.transform(BodyTransformer.java:51)
	at soot.Transform.apply(Transform.java:120)
	at soot.JimpleBodyPack.applyPhaseOptions(JimpleBodyPack.java:66)
	at soot.JimpleBodyPack.internalApply(JimpleBodyPack.java:105)
	at soot.Pack.apply(Pack.java:125)
	at soot.asm.AsmMethodSource.getBody(AsmMethodSource.java:2066)
	... 5 more
Exception in thread "Thread-26" java.lang.RuntimeException: Failed to apply jb to <org.springframework.core.io.buffer.DataBufferUtils$WritableByteChannelSubscriber: void hookOnNext(org.springframework.core.io.buffer.DataBuffer)>
	at soot.asm.AsmMethodSource.getBody(AsmMethodSource.java:2068)
	at soot.SootMethod.retrieveActiveBody(SootMethod.java:402)
	at soot.PackManager$1.run(PackManager.java:1279)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: no active FastHierarchy present for scene
	at soot.Scene.getFastHierarchy(Scene.java:1296)
	at soot.Scene.getOrMakeFastHierarchy(Scene.java:1287)
	at soot.toolkits.exceptions.ThrowableSet.whichCatchableAs(ThrowableSet.java:705)
	at soot.toolkits.graph.ExceptionalUnitGraph.buildExceptionDests(ExceptionalUnitGraph.java:290)
	at soot.toolkits.graph.ExceptionalUnitGraph.initialize(ExceptionalUnitGraph.java:229)
	at soot.toolkits.graph.ExceptionalUnitGraph.<init>(ExceptionalUnitGraph.java:135)
	at soot.toolkits.graph.ExceptionalUnitGraph.<init>(ExceptionalUnitGraph.java:163)
	at soot.jimple.toolkits.base.Aggregator.internalAggregate(Aggregator.java:129)
	at soot.jimple.toolkits.base.Aggregator.internalTransform(Aggregator.java:114)
	at soot.BodyTransformer.transform(BodyTransformer.java:51)
	at soot.Transform.apply(Transform.java:120)
	at soot.JimpleBodyPack.applyPhaseOptions(JimpleBodyPack.java:66)
	at soot.JimpleBodyPack.internalApply(JimpleBodyPack.java:105)
	at soot.Pack.apply(Pack.java:125)
	at soot.asm.AsmMethodSource.getBody(AsmMethodSource.java:2066)
	... 5 more
java.lang.RuntimeException: Failed to apply jb to <org.springframework.core.io.buffer.DataBufferUtils$WritableByteChannelSubscriber: void hookOnNext(org.springframework.core.io.buffer.DataBuffer)>
	at soot.asm.AsmMethodSource.getBody(AsmMethodSource.java:2068)
	at soot.SootMethod.retrieveActiveBody(SootMethod.java:402)
	at soot.PackManager$1.run(PackManager.java:1279)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: no active FastHierarchy present for scene
	at soot.Scene.getFastHierarchy(Scene.java:1296)
	at soot.Scene.getOrMakeFastHierarchy(Scene.java:1287)
	at soot.toolkits.exceptions.ThrowableSet.whichCatchableAs(ThrowableSet.java:705)
	at soot.toolkits.graph.ExceptionalUnitGraph.buildExceptionDests(ExceptionalUnitGraph.java:290)
	at soot.toolkits.graph.ExceptionalUnitGraph.initialize(ExceptionalUnitGraph.java:229)
	at soot.toolkits.graph.ExceptionalUnitGraph.<init>(ExceptionalUnitGraph.java:135)
	at soot.toolkits.graph.ExceptionalUnitGraph.<init>(ExceptionalUnitGraph.java:163)
	at soot.jimple.toolkits.base.Aggregator.internalAggregate(Aggregator.java:129)
	at soot.jimple.toolkits.base.Aggregator.internalTransform(Aggregator.java:114)
	at soot.BodyTransformer.transform(BodyTransformer.java:51)
	at soot.Transform.apply(Transform.java:120)
	at soot.JimpleBodyPack.applyPhaseOptions(JimpleBodyPack.java:66)
	at soot.JimpleBodyPack.internalApply(JimpleBodyPack.java:105)
	at soot.Pack.apply(Pack.java:125)
	at soot.asm.AsmMethodSource.getBody(AsmMethodSource.java:2066)
	... 5 more
```

**Comment by mbenz89** [comment]
Hi Nico.

Thanks for investigating and the in-depth report!
I just pushed a [quick fix](a82c2a96ccfbdf358edf76ff73ef5081b11e216b) for the issue. Could you please check if that solves your problem?

The problem is that Soot wasn't really built with parallelization in mind. It was added much later but not all code was adapted accordingly. The reported error was caused by not properly synchronizing writes/reads of the `FastHierarchy` in the `Scene`, as the exception rather clearly shows.

Best,
Manuel

**Comment by nrosner** [comment]
Manuel, thank you very much for the quick fix!
